### ISSUE TITLE
Add support of timestamped value for Observe-Composite.

### ISSUE THREAD

**Issue by sbernard31** [eclipse/leshan#1089]
This feature is needed to support clients which are using `Notification Storing When Disabled or Offline` feature. 

Currently,  
Leshan client does not support this feature (see #596)  
Leshan server is able to handle "classic" observe response from client using this feature.  
Leshan server is not able to handle "composite" observe response from client using this feature.

**Comment by JaroslawLegierski** [comment]
I prepared following [PoC](https://github.com/eclipse-leshan/leshan/compare/master...JaroslawLegierski:leshan:opl/observe_composite_support_for_ts_values) with very limited support for Observe-Composite (only in leshan client) with timestamped data 
(mainly for better understunding observe-composite concept in Leshan)

What do you think about it - is this the correct direction? 
Currently we have 2 different methods in encoder:

-  encode - for classical nodes
-  encodeTimestampedNode - for timestamped nodes

Do we need 3'rd encoder for encoding mixed nodes (timestamped with classical into one notification) ?

**Comment by sbernard31** [comment]
I looked at it and don't think this is the right direction.

1. There is no need to code new encoder/decoder `TimestampedMultiNodeEncoder`/`TimestampedMultiNodeDecoder` should be enough
2. Maybe `TimestampedMultiNodeDecoder.decodeTimestampedNodes(byte[], LwM2mModel)` should be adapted as expected (https://github.com/eclipse-leshan/leshan/pull/1218#discussion_r824550092)
3. No need to add client code to test it, just create a dedicated integration tests like `ObserveTimeStampTest`
4. (not directly linked I guess)  I didn't get why you add `.getId(`) to `LwM2mRoot()`  ?

HTH :slightly_smiling_face:

**Comment by sbernard31** [comment]
> OK then I will focus on adapting TimestampedMultiNodeDecoder.decodeTimestampedNodes(byte[], LwM2mModel)

Let me know, if there is unclear point. 

> As I understand it, I should start by writing dedicated interation test instead of modifying the client/leshan-client-demo code ?

Yep I think this will be easier and adding a integration tests is a good idea anyway.

**Comment by JaroslawLegierski** [comment]
Please find [here ](https://github.com/eclipse-leshan/leshan/compare/master...JaroslawLegierski:leshan:opl/observe_composite_timestamped_data_support_for_server) PoC of Observe Composite  with ts values support on server side.

The following points are not clear for me:

1) Currently leshan SENML_JSON encoder can encode following  records:

```json
[
	{
		"bn": "/3303/0/5700",
		"bt": 1699877805.766,
		"v": -128.8
	},
	{
		"bn": "/6/0/0",
		"bt": 1699877805.800,
		"v": -39.0
	}
]
```
can we change this format to [following](https://github.com/eclipse-leshan/leshan/compare/master...JaroslawLegierski:leshan:opl/observe_composite_support_for_ts_values#diff-12f89c2d13267e7612a28f71c3e1e8385b0dfd83f72cb56e7d89b8adf76cb956) one: ?
```json
[
	{
		"bn": "/3303/0/5700",
		"bt": 1699877805.766,
		"v": -128.8
	},
	{
		"bn": "/6/0/0",
		"t": 0.101,
		"v": -39.0
	}
]
```
2)  Support of mixed data ([Multiple Measurements](https://datatracker.ietf.org/doc/html/rfc8428#section-5.1.3))  when we have data with and without timestamps in one record e.g.:

```json
[
	{
		"bn": "/3303/0/5700",
		"bt": 1699877805.766,
		"v": -128.8
	},
	{
		"bn": "/6/0/0",
		"v": -39.0
	}
]
```
Writing test that produces such data requires modifying the encoder (on the client side). Should I modify the existing encoder or create dedicated one only to be used by the integration test ?

**Comment by sbernard31** [comment]
I looked at the POC quickly, I think this is the right direction. You can create a PR. 

About SENML encoding, this is not related to this issue, so better to create a dedicated one.

**Comment by JaroslawLegierski** [comment]
> E.g. : still with `/1/0/1` and `/3` , the notification should contains contains both nodes for each timestamp.
> For TimestampedLwM2mNodes, I think this should looks like :

As I understand in this case correct SenML format is as follow ?

```json
[
	{
		"bn": "/1/0/1",
		"bt": 1701940138.352,
		"v": 3600
	},
	{
		"bn": "/3/",
		"bt": 1701940138.352,
		"n": "0/0",
		"vs": "Leshan Demo Device"
	},
	{
		"n": "0/1",
		"vs": "Model 500"
	},
	{
		"n": "0/2",
		"vs": "LT-500-000-0001"
	},
	{
		"n": "0/3",
		"vs": "1.0.0"
	},
	{
		"n": "0/9",
		"v": 7
	},
	{
		"n": "0/10",
		"v": 276427
	},
	{
		"n": "0/11/0",
		"v": 0
	},
	{
		"n": "0/13",
		"v": 1701937693
	},
	{
		"n": "0/14",
		"vs": "+01"
	},
	{
		"n": "0/15",
		"vs": "Europe/Belgrade"
	},
	{
		"n": "0/16",
		"vs": "U"
	},
	{
		"n": "0/17",
		"vs": "Demo"
	},
	{
		"n": "0/18",
		"vs": "1.0.1"
	},
	{
		"n": "0/19",
		"vs": "1.0.2"
	},
	{
		"n": "0/20",
		"v": 2
	},
	{
		"n": "0/21",
		"v": 296960
	}
]
```

**Comment by JaroslawLegierski** [comment]
> So do we agree on : [#1089 (comment)](https://github.com/eclipse-leshan/leshan/issues/1089#issuecomment-1842607310) 🙂 ?

Yes, I agree - I'm currently working on correcting the integration tests

**Comment by JaroslawLegierski** [comment]
I just finished correction of the integration tests. PR #1552  has been updated.

**Comment by sbernard31** [comment]
I will look at https://github.com/eclipse-leshan/leshan/pull/1552 and test it and probably try to modify the code a bit.
